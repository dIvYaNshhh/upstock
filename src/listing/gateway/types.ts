export type HoldingItem = {
  symbol: string
  quantity: number
  ltp: number
  avgPrice: number
  close: number
}

export type Data = {
  userHolding: HoldingItem[]
}

export type HoldingResponse = {
  data: Data
}
