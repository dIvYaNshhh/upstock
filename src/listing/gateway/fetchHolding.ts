import { HoldingItem, HoldingResponse } from './types'

const holdingUrl = 'https://35dee773a9ec441e9f38d5fc249406ce.api.mockbin.io/'

export const fetchHoldiing = async (): Promise<HoldingResult> => {
  try {
    const response = await fetch(holdingUrl)
    if (response.ok || response.status == 200) {
      const holdingResponse: HoldingResponse = await response.json()
      return {
        holding: holdingResponse.data.userHolding,
        status: true,
      }
    } else {
      return {
        holding: undefined,
        status: false,
      }
    }
  } catch (error) {
    console.log(error)
    return {
      holding: undefined,
      status: false,
    }
  }
}

export type HoldingResult = {
  holding: HoldingItem[] | undefined
  status: boolean | undefined
}
