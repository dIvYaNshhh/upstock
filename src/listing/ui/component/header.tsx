import React from 'react'
import { View, Text } from 'react-native'
import styled from 'styled-components'

type Props = {
  title: string
}
export const Header = ({ title }: Props) => {
  return (
    <Container>
      <Title>{title}</Title>
    </Container>
  )
}

const Container = styled(View)`
  background-color: #7a007a;
`

const Title = styled(Text)`
  color: #fff;
  font-size: 20px;
  padding: 16px;
`
