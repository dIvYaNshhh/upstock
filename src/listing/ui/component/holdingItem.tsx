import React, { useState } from 'react'
import { Text, View, TouchableOpacity } from 'react-native'
import styled from 'styled-components'
import Collapsible from 'react-native-collapsible'
import { HoldingItem } from 'src/listing/gateway/types'
import { usePortfolioData } from 'src/listing/bussinessLogic/usePortfolioData'

interface Props {
  item: HoldingItem
}
const currencySymbol = 'â‚¹'
export const HoldingListItem = React.memo(({ item }: Props) => {
  const { currentValue, investmentValue, pnl, todaysPnl } = usePortfolioData({
    item,
  })

  const [collapsed, setCollapsed] = useState(true)

  const toggleExpanded = () => {
    setCollapsed(!collapsed)
  }

  return (
    <View>
      <Container onPress={toggleExpanded}>
        <Column flexEnd={false}>
          <Symbol>{item.symbol}</Symbol>
          <TextView>{item.quantity}</TextView>
        </Column>

        <Column flexEnd={true}>
          <TextView>
            {'LTP: '}
            <MildBoldText>{`${currencySymbol} ${item.ltp}`}</MildBoldText>
          </TextView>
          <TextView>
            {'P/L: '}
            <MildBoldText>{`${currencySymbol} ${todaysPnl.toFixed(
              2
            )}`}</MildBoldText>
          </TextView>
        </Column>
      </Container>
      <Collapsible collapsed={collapsed}>
        <ExpandedSection>
          <Row>
            <MildBoldText>{'Current Value: '}</MildBoldText>
            <TextView>{`${currencySymbol} ${currentValue.toFixed(
              2
            )}`}</TextView>
          </Row>
          <Row>
            <MildBoldText>{'Total Investment: '}</MildBoldText>
            <TextView>{`${currencySymbol} ${investmentValue.toFixed(
              2
            )}`}</TextView>
          </Row>
          <Row>
            <MildBoldText>{`Today's Profit & Loss: `}</MildBoldText>
            <TextView>{`${currencySymbol} ${pnl.toFixed(2)}`}</TextView>
          </Row>
          <Space />
          <Row>
            <MildBoldText>{'Profit & Loss: '}</MildBoldText>
            <TextView>{`${currencySymbol} ${pnl.toFixed(2)}`}</TextView>
          </Row>
        </ExpandedSection>
      </Collapsible>
    </View>
  )
})

const Container = styled(TouchableOpacity)`
  padding: 16px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
`
const Column = styled(View)<{
  flexEnd: boolean
}>`
  flex: 1;
  flex-direction: column;
  align-items: ${({ flexEnd }) => (flexEnd ? 'flex-end' : 'flex-start')};
`

const Symbol = styled(Text)`
  font-weight: bold;
  font-size: 16px;
  padding: 2px;
`
const TextView = styled(Text)`
  font-size: 16px;
  padding: 2px;
`

const MildBoldText = styled(Text)`
  font-weight: bold;
  font-size: 16px;
  padding: 2px;
`
const ExpandedSection = styled(View)`
  padding: 16px;
`
const Row = styled(View)`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
`

const Space = styled(View)`
  height: 16px;
`
