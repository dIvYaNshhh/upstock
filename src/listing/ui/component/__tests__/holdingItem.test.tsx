import React from 'react'
import { render, RenderAPI } from 'react-native-testing-library'
import { HoldingItem } from 'src/listing/gateway/types'
import { HoldingListItem } from 'src/listing/ui/component/holdingItem'

const item: HoldingItem = {
  avgPrice: 100,
  close: 100,
  ltp: 100,

  symbol: 'AAPL',
  quantity: 100,
}
describe('Component - HoldingListItem', () => {
  let instance: RenderAPI

  beforeEach(() => {
    instance = render(<HoldingListItem key={0} item={item} />)
  })

  it('matches the snapshot', () => {
    expect(instance.toJSON()).toMatchSnapshot()
  })
})
