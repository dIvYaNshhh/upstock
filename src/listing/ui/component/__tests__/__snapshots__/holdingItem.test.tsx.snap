// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - HoldingListItem matches the snapshot 1`] = `
<View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      className="sc-bwzfXH cWdhba"
      flexEnd={false}
    >
      <Text
        className="sc-htpNat cjXytN"
      >
        AAPL
      </Text>
      <Text
        className="sc-bxivhb esjODI"
      >
        100
      </Text>
    </View>
    <View
      className="sc-bwzfXH fjyMey"
      flexEnd={true}
    >
      <Text
        className="sc-bxivhb esjODI"
      >
        LTP: 
        <Text
          className="sc-ifAKCX kJknWg"
        >
          ₹ 100
        </Text>
      </Text>
      <Text
        className="sc-bxivhb esjODI"
      >
        P/L: 
        <Text
          className="sc-ifAKCX kJknWg"
        >
          ₹ 0.00
        </Text>
      </Text>
    </View>
  </View>
  <View
    collapsable={false}
    pointerEvents="none"
    style={
      {
        "height": 0,
        "overflow": "hidden",
      }
    }
  >
    <View
      collapsable={false}
      onLayout={[Function]}
      style={{}}
    >
      <View
        className="sc-EHOje gDoXQW"
      >
        <View
          className="sc-bZQynM jbVsDv"
        >
          <Text
            className="sc-ifAKCX kJknWg"
          >
            Current Value: 
          </Text>
          <Text
            className="sc-bxivhb esjODI"
          >
            ₹ 10000.00
          </Text>
        </View>
        <View
          className="sc-bZQynM jbVsDv"
        >
          <Text
            className="sc-ifAKCX kJknWg"
          >
            Total Investment: 
          </Text>
          <Text
            className="sc-bxivhb esjODI"
          >
            ₹ 10000.00
          </Text>
        </View>
        <View
          className="sc-bZQynM jbVsDv"
        >
          <Text
            className="sc-ifAKCX kJknWg"
          >
            Today's Profit & Loss: 
          </Text>
          <Text
            className="sc-bxivhb esjODI"
          >
            ₹ 0.00
          </Text>
        </View>
        <View
          className="sc-gzVnrw bqaexq"
        />
        <View
          className="sc-bZQynM jbVsDv"
        >
          <Text
            className="sc-ifAKCX kJknWg"
          >
            Profit & Loss: 
          </Text>
          <Text
            className="sc-bxivhb esjODI"
          >
            ₹ 0.00
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
